### DISCUSSION TITLE
Not a technical question

### DISCUSSION THREAD

**Discussion by GINNMAX** [MREYE-LUMC/ZOSPy#50] [Q&A]
I'm a master student in lens and I have used the ZOSpy, Its python style and is away from cpp style, i also interested in coding. iwant to ask dou you have CS backgroud or just learn by yourself

**Comment by crnh** [comment]
Hi @GINNMAX, thanks for your question! I have a background in physics and medical physics. Prior to my studies, I learned programming (in C# / .NET) at a side job in a software company. There I acquired most basic programming skills, which I further developed during my studies and later side jobs. I learnt Python mostly by myself, but my corporate experience with C# definitely helped with that.

If programming is not in your curriculum, I recommend to follow a few (free) online Python courses to get the basics right. If you just learn Python by trial and error, you're likely to develop some bad habits, so following these courses is important to get aware of the best practices. [Codecademy](https://www.codecademy.com/) offers some good courses, and I also have good experiences with [SoloLearn](https://www.sololearn.com/) (when I did their Python course it was much less interactive than the course offered by codecademy, but it paid more attention to more advanced topics, which was nice). I also strongly recommend [Exercism](https://exercism.org/), which is an online platform where you can solve coding challenges and then get feedback about the quality of your solutions. Not many platforms offer this for free, and their mentoring option is definitely worth a try!

Last but not least, as you are a master's student, you are probably eligible for the [GitHub Student Developer Pack](https://education.github.com/pack). If you sign up for this program, you will get free access to some development tools, free GitHub Pro, and free access to some online learning platforms (e.g. Educative, Datacamp, One Month, and PluralSight through Microsoft Visual Studio Dev Essentials).

The backgrounds of the other ZOSPy developers are rather diverse. @LucVV how did you get into programming?

**Comment by LucVV** [comment]
Myself I started coding during my PhD. My background is in medicine, so I had to start with the basics! Due to the nature of my PhD, it required or would at least benefit greatly from automation of OpticStudio.

I did some online courses in Python to understand the basics, mostly on Datacamp, and asked colleagus that were proficient in coding for explanations where needed. I started by crwating some scripts that used PyzDDE l, but as the DDE of OpticStudio became depricated, I started interacting with the API. To make my code more concise (and clear for my colleagues), I started making classes and functions, which we eventually turned into ZOSPy. Long story short: it was a gradual process and I am very greatfull for all the help that I had along the way. I am still learning a lot, and I guess that will be the case for the foreseeable future.

What helped me a lot is to take some small projects at first. First, I automated some small things for local projects with some scripts. After I had some more understanding of Python, I made a small pull request for PyzDDE with just a small addition, and the feedback on that request gave me more insight in how things work. 

Hope it helps!
